Data/ByteString/Char8.hs

```diff
@@ -1001,21 +1005,134 @@ unwords = intercalate (singleton ' ')
 readInt :: ByteString -> Maybe (Int, ByteString)
 {-# INLINABLE readInt #-}
 readInt bs = case B.uncons bs of
-    Just (w, rest) | w - 0x30 <= 9 -> readDec True bs    -- starts with digit
-                   | w == 0x2d     -> readDec False rest -- starts with minus
-                   | w == 0x2b     -> readDec True rest  -- starts with plus
-    _                              -> Nothing            -- not signed decimal
+    Just (w, rest) | w - 0x30 <= 9 -> cvtpos $ readWord64 bs
+                   | w == 0x2d     -> cvtneg $ readWord64 rest
+                   | w == 0x2b     -> cvtpos $ readWord64 rest
+    _                              -> Nothing
   where
-    -- | Read a decimal 'Int' without overflow.  The caller has already
+    -- Conversion from Word64 to a positive Int is value-preserving provided
+    -- the input does not exceed @maxBound :: Int@ as a 64-bit value.
+    cvtpos (Just (w, str))
+        | w <= intMaxAsWord64
+             = let !i = fromIntegral w in Just (i, str)
+    cvtpos _ = Nothing
+
+    -- We accept '-0' as well as '+0'.  Conversion from Word64 to a negative
+    -- Int is value-preserving provided the input does not exceed @'maxBound'
+    -- :: 'Word'@, and the sign of the result is negative.
+    --
+    cvtneg (Just (w, str))
+        | w <= wordMaxAsWord64
+        , let !i = negate $ fromIntegral w
+        , i <= 0 = Just (i, str)
```

Comment:
This could probably also use `w <= negatedIntMinAsWord64` to save a comparison.

Reply:
How do you cleanly express `negatedIntMinAsWord64`?  The naive `fromIntegral (minBound :: Int) :: Word64` isn't it...

Reply:
My less naive version is: `fromIntegral (fromIntegral (minBound :: Int) :: Word) :: Word64`

Reply:
Nice :)

Reply:
Which, with`TypeApplications` I guess becomes:
```
intMinAsWord64  = fromIntegral @Word @Word64 $ fromIntegral @Int @Word minBound 
```
Do you want to force "neg" into the name, or is it OK to leave it implicit that this is an absolute value?

Reply:
Up to you.